File: client/lib/models/study_pack.dart
class StudySourceRef {
  final int segmentIndex;
  final String? speaker;
  final double? start;
  final double? end;
  final String text;

  const StudySourceRef({
    required this.segmentIndex,
    this.speaker,
    this.start,
    this.end,
    required this.text,
  });

  factory StudySourceRef.fromJson(Map<String, dynamic> json) => StudySourceRef(
        segmentIndex: json['segment_index'] as int,
        speaker: json['speaker'] as String?,
        start: (json['start'] as num?)?.toDouble(),
        end: (json['end'] as num?)?.toDouble(),
        text: json['text'] as String? ?? '',
      );
}

class StudyKeyConcept {
  final String term;
  final String explanation;
  final List<int> sourceRefs;

  const StudyKeyConcept({
    required this.term,
    required this.explanation,
    this.sourceRefs = const [],
  });

  factory StudyKeyConcept.fromJson(Map<String, dynamic> json) =>
      StudyKeyConcept(
        term: json['term'] as String? ?? '',
        explanation: json['explanation'] as String? ?? '',
        sourceRefs: _parseIntList(json['source_refs']),
      );
}

class StudyFlashcard {
  final String front;
  final String back;
  final List<int> sourceRefs;

  const StudyFlashcard({
    required this.front,
    required this.back,
    this.sourceRefs = const [],
  });

  factory StudyFlashcard.fromJson(Map<String, dynamic> json) => StudyFlashcard(
        front: json['front'] as String? ?? '',
        back: json['back'] as String? ?? '',
        sourceRefs: _parseIntList(json['source_refs']),
      );
}

class StudyQuizQuestion {
  final String question;
  final String answer;
  final String difficulty;
  final List<int> sourceRefs;

  const StudyQuizQuestion({
    required this.question,
    required this.answer,
    required this.difficulty,
    this.sourceRefs = const [],
  });

  factory StudyQuizQuestion.fromJson(Map<String, dynamic> json) =>
      StudyQuizQuestion(
        question: json['question'] as String? ?? '',
        answer: json['answer'] as String? ?? '',
        difficulty: json['difficulty'] as String? ?? 'medium',
        sourceRefs: _parseIntList(json['source_refs']),
      );
}

class StudyPack {
  final String taskId;
  final String mode;
  final String language;
  final List<StudyKeyConcept> keyConcepts;
  final List<StudyFlashcard> flashcards;
  final List<StudyQuizQuestion> quizQuestions;
  final String studyNotes;
  final List<StudySourceRef> sourceRefs;
  final String createdAt;

  const StudyPack({
    required this.taskId,
    required this.mode,
    required this.language,
    required this.keyConcepts,
    required this.flashcards,
    required this.quizQuestions,
    required this.studyNotes,
    required this.sourceRefs,
    required this.createdAt,
  });

  factory StudyPack.fromJson(Map<String, dynamic> json) => StudyPack(
        taskId: json['task_id'] as String? ?? '',
        mode: json['mode'] as String? ?? 'general',
        language: json['language'] as String? ?? 'ko',
        keyConcepts: (json['key_concepts'] as List<dynamic>? ?? [])
            .whereType<Map<String, dynamic>>()
            .map(StudyKeyConcept.fromJson)
            .toList(),
        flashcards: (json['flashcards'] as List<dynamic>? ?? [])
            .whereType<Map<String, dynamic>>()
            .map(StudyFlashcard.fromJson)
            .toList(),
        quizQuestions: (json['quiz_questions'] as List<dynamic>? ?? [])
            .whereType<Map<String, dynamic>>()
            .map(StudyQuizQuestion.fromJson)
            .toList(),
        studyNotes: json['study_notes'] as String? ?? '',
        sourceRefs: (json['source_refs'] as List<dynamic>? ?? [])
            .whereType<Map<String, dynamic>>()
            .map(StudySourceRef.fromJson)
            .toList(),
        createdAt: json['created_at'] as String? ?? '',
      );
}

List<int> _parseIntList(dynamic value) {
  if (value is! List) return [];
  return value
      .where((item) => item is int || item is num)
      .map((item) => (item as num).toInt())
      .toList();
}

File: backend/schemas/study_pack.py
"""Study Pack schemas for transcript-based learning artifacts."""

from enum import StrEnum

from pydantic import BaseModel, Field


class StudyPackMode(StrEnum):
    """Learning artifact mode."""

    GENERAL = "general"
    LECTURE = "lecture"
    MEETING = "meeting"
    INTERVIEW = "interview"
    SERMON = "sermon"


class StudyPackCreateRequest(BaseModel):
    """Study pack generation request."""

    mode: StudyPackMode = Field(default=StudyPackMode.GENERAL, description="Study pack mode")
    language: str = Field(default="ko", min_length=2, max_length=16, description="Output language")
    max_tokens: int = Field(
        default=1800,
        ge=512,
        le=4096,
        description="ZAI-compatible API max response tokens",
    )
    force_refresh: bool = Field(default=False, description="Regenerate even when cached")


class StudySourceRef(BaseModel):
    """Source transcript reference."""

    segment_index: int = Field(ge=0, description="Transcript segment index")
    speaker: str | None = Field(default=None, description="Speaker label")
    start: float | None = Field(default=None, ge=0.0, description="Segment start seconds")
    end: float | None = Field(default=None, ge=0.0, description="Segment end seconds")
    text: str = Field(default="", description="Source text excerpt")


class StudyKeyConcept(BaseModel):
    """Key concept extracted from a transcript."""

    term: str = Field(..., min_length=1, description="Concept term")
    explanation: str = Field(..., min_length=1, description="Grounded explanation")
    source_refs: list[int] = Field(default_factory=list, description="Segment indexes")


class StudyFlashcard(BaseModel):
    """Flashcard item."""

    front: str = Field(..., min_length=1, description="Prompt side")
    back: str = Field(..., min_length=1, description="Answer side")
    source_refs: list[int] = Field(default_factory=list, description="Segment indexes")


class StudyQuizQuestion(BaseModel):
    """Quiz question item."""

    question: str = Field(..., min_length=1, description="Quiz question")
    answer: str = Field(..., min_length=1, description="Correct answer")
    difficulty: str = Field(default="medium", description="easy/medium/hard")
    source_refs: list[int] = Field(default_factory=list, description="Segment indexes")


class StudyPackResponse(BaseModel):
    """Generated study pack response."""

    task_id: str = Field(..., min_length=1, description="Source minutes task ID")
    mode: StudyPackMode = Field(default=StudyPackMode.GENERAL)
    language: str = Field(default="ko")
    key_concepts: list[StudyKeyConcept] = Field(default_factory=list)
    flashcards: list[StudyFlashcard] = Field(default_factory=list)
    quiz_questions: list[StudyQuizQuestion] = Field(default_factory=list)
    study_notes: str = Field(default="")
    source_refs: list[StudySourceRef] = Field(default_factory=list)
    created_at: str = Field(..., description="ISO-8601 creation timestamp")
